examples/1d_stencil/1d_stencil_4_repart.cpp

```diff
@@ -0,0 +1,381 @@
+//  Copyright (c) 2014 Hartmut Kaiser
+//  Copyright (c) 2014 Patricia Grubel
+//  Copyright (c) 2015 Oregon University
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// This is the fourth in a series of examples demonstrating the development of
+// a fully distributed solver for a simple 1D heat distribution problem.
+//
+// This example builds on example three. It futurizes the code from that
+// example. Compared to example two this code runs much more efficiently. It
+// allows for changing the amount of work executed in one HPX thread which
+// enables tuning the performance for the optimal grain size of the
+// computation. This example is still fully local but demonstrates nice
+// scalability on SMP machines.
+
+#include <hpx/hpx_init.hpp>
+#include <hpx/hpx.hpp>
+
+#include <hpx/include/parallel_algorithm.hpp>
+#include <hpx/include/performance_counters.hpp>
+
+#include <boost/range/irange.hpp>
+#include <boost/cstdint.hpp>
+#include <boost/format.hpp>
+
+#include <limits>
+#include <algorithm>
+
+#include <apex.hpp>
+#include <apex_api.hpp>
+
+#include "print_time_results.hpp"
+
+using hpx::naming::id_type;
+using hpx::performance_counters::get_counter;
+using hpx::performance_counters::stubs::performance_counter;
+using hpx::performance_counters::counter_value;
+using hpx::performance_counters::status_is_valid;
+
+static bool counters_initialized = false;
+static const char * counter_name = "/threads{locality#%d/total}/idle-rate";
+static apex_event_type end_iteration_event = APEX_CUSTOM_EVENT;
+static hpx::naming::id_type counter_id;
+
+
+id_type get_counter_id() {
+    // Resolve the GID of the performances counter using it's symbolic name.
+    boost::uint32_t const prefix = hpx::get_locality_id();
+    boost::format active_threads(counter_name);
+    id_type id = get_counter(boost::str(active_threads % prefix));
+    return id;
+}
+
+void setup_counters() {
+    try {
+        id_type id = get_counter_id();
+        // We need to explicitly start all counters before we can use them. For
+        // certain counters this could be a no-op, in which case start will return
+        // 'false'.
+        performance_counter::start(id);
+        std::cout << "Counters initialized! " << id << std::endl;
+        counter_value value = performance_counter::get_value(id);
+        std::cout << "Idle Rate " << value.get_value<boost::int64_t>() << std::endl;
+        counter_id = id;
+        end_iteration_event = apex::register_custom_event("Repartition");
+    }
+    catch(hpx::exception const& e) {
+        std::cerr << "apex_policy_engine_active_thread_count: caught exception: "
+            << e.what() << std::endl;
```

Comment:
In the case of an exception, I think this function should return and counter_id should be set to an invalid id to avoid further problems.